### PR TITLE
Use correct master url for token requests, set master CA correctly

### PR DESCRIPTION
**Pull request by liggitt**
Fixes https://github.com/openshift/origin/issues/4894

### CODE REVIEW

**Comment by liggitt** [comment]
@deads ptal

**Comment by liggitt** [comment]
[test]

File: pkg/cmd/server/api/types.go [hunk 1]
```diff
@@ -418,7 +421,7 @@ type ServiceAccountConfig struct {
 
 	// MasterCA is the CA for verifying the TLS connection back to the master.  The service account controller will automatically
 	// inject the contents of this file into pods so they can verify connections to the master.
-	MasterCA string
+	MasterCA *string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: pretty sure we should have done this as a pointer originally, to distinguish between missing, intentionally set to nothing, and present

Comment (deads2k) [inline pkg/cmd/server/api/types.go]: > pretty sure we should have done this as a pointer originally, to distinguish between missing, intentionally set to nothing, and present

Why is the distinction between missing and intentionally set to nothing important in this case?

Comment (liggitt) [inline pkg/cmd/server/api/types.go]: in case they want to use system roots

Comment (deads2k) [inline pkg/cmd/server/api/types.go]: > in case they want to use system roots. missing has to migrate the value we were previously using

In what case would I want to specifically exclude the system roots?  If that's the only reason to have this distinction, then its not very valuable.

**Comment by liggitt** [comment]
reverted service account masterCA changes

File: pkg/cmd/server/api/v1/types_test.go [hunk 2]
```diff
@@ -294,9 +299,12 @@ func TestMasterConfig(t *testing.T) {
 				{Provider: runtime.EmbeddedObject{Object: &internal.OpenIDIdentityProvider{}}},
 			},
 			SessionConfig: &internal.SessionConfig{},
+			Templates:     &internal.OAuthTemplates{},
+		},
+		AssetConfig: &internal.AssetConfig{
```
Comment (deads2k) [inline pkg/cmd/server/api/v1/types_test.go]: I haven't seen this before.  Reviewed in some other pull?

File: pkg/cmd/server/api/validation/oauth.go [hunk 3]
```diff
@@ -18,8 +18,16 @@ import (
 func ValidateOAuthConfig(config *api.OAuthConfig) ValidationResults {
 	validationResults := ValidationResults{}
 
+	if config.MasterCA != nil && len(*config.MasterCA) > 0 {
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/oauth.go]: post-conversion, this is never nil, right?  Clayton declared no nil checking.

File: pkg/cmd/server/api/validation/oauth.go [hunk 4]
```diff
@@ -18,8 +18,16 @@ import (
...
+		validationResults.AddErrors(ValidateFile(*config.MasterCA, "masterCA")...)
+	} else {
+		validationResults.AddWarnings(fielderrors.NewFieldInvalid("masterCA", "", "connections to the masterURL for OAuth token exchange will use system roots"))
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/oauth.go]: Why is this worthy of a warning?  You think it would be unexpected in some way?

File: pkg/cmd/server/api/validation/oauth.go [hunk 5]
```diff
@@ -18,8 +18,16 @@ import (
...
+	}
+
 	if len(config.MasterURL) == 0 {
 		validationResults.AddErrors(fielderrors.NewFieldRequired("masterURL"))
+	} else if _, urlErrs := ValidateURL(config.MasterURL, "masterURL"); len(urlErrs) > 0 {
+		validationResults.AddErrors(urlErrs...)
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/oauth.go]: Shouldn't you just be able to add the `urlErrs` and have the `AddErrors` correctly handle an empty slice?

File: pkg/cmd/server/origin/auth.go [hunk 6]
```diff
@@ -152,12 +154,23 @@ func (c *AuthConfig) InstallAPI(container *restful.Container) []string {
...
+	if c.Options.MasterCA != nil && len(*c.Options.MasterCA) > 0 {
+		rootCAs, err := cmdutil.CertPoolFromFile(*c.Options.MasterCA)
+		if err != nil {
+			glog.Fatal(err)
+		}
+
+		osOAuthClient.Transport = &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			Dial: (&net.Dialer{
+				Timeout:   30 * time.Second,
```
Comment (deads2k) [inline pkg/cmd/server/origin/auth.go]: Why are timeouts only set if I have a `MasterCA` and not when I run with system defaults.

File: pkg/cmd/server/origin/auth.go [hunk 7]
```diff
@@ -152,12 +154,23 @@ func (c *AuthConfig) InstallAPI(container *restful.Container) []string {
...
-		// Set TLS CA roots
-		transport.TLSClientConfig = &tls.Config{RootCAs: c.MasterRoots}
-		osOAuthClient.Transport = &transport
+	if c.Options.MasterCA != nil && len(*c.Options.MasterCA) > 0 {
+		rootCAs, err := cmdutil.CertPoolFromFile(*c.Options.MasterCA)
+		if err != nil {
+			glog.Fatal(err)
+		}
+
+		osOAuthClient.Transport = &http.Transport{
```
Comment (deads2k) [inline pkg/cmd/server/origin/auth.go]: its not obvious why the changes to this structure unrelated to the `rootCA` are still in this `if`.

Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: this is identical to the default transport... I'm pretty sure it was improper to create a copy of it and override it previously...

File: pkg/cmd/server/origin/auth.go (hunk 6 above)
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: this is identical to the default transport that gets used if osOAuthClient.Transport isn't set, with a custom TLSClientConfig

File: pkg/cmd/server/api/v1/types_test.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/server/api/v1/types_test.go]: Yeah, that was the UI extension configuration, I wanted the substructs exposed in the unit test

File: pkg/cmd/server/origin/auth.go (hunk 7 above)
Comment (deads2k) [inline pkg/cmd/server/origin/auth.go]: > this is identical to the default transport... I'm pretty sure it was improper to create a copy of it and override it previously...

Definitely needs a comment indicating that the intent is to reproduce the defaults.  It's not obvious.

File: pkg/cmd/server/start/master_args.go [hunk 8]
```diff
@@ -256,6 +255,11 @@ func (args MasterArgs) BuildSerializeableMasterConfig() (*configapi.MasterConfig
 
 		config.AssetConfig.ServingInfo.ServerCert = admin.DefaultAssetServingCertInfo(args.ConfigDir.Value())
 
+		if oauthConfig != nil {
+			s := admin.DefaultRootCAFile(args.ConfigDir.Value())
```
Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: Why not pull the value from the master serving info?

Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: > Why not pull the value from the master serving info?

Oh, right.  Nevermind.

File: pkg/cmd/server/api/validation/oauth.go (hunk 3 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: moved the defaulting for this to a defaulting func on the internal version and removed the nil check

File: pkg/cmd/server/api/validation/oauth.go (hunk 4 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: removed warning

File: pkg/cmd/server/api/validation/oauth.go (hunk 5 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: this keeps the structure clean if we wanted to chain more checks after this if block for a clean url. it also matches masterPublicURL handling

File: pkg/cmd/server/origin/auth.go (hunk 7 above)
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: pulled in a defaulting func for transports from upstream and cleaned up some places we were constructing transports. now the only bit here is the part we're interested in... the rest comes from the defaults

File: pkg/cmd/server/origin/auth.go [hunk 9]
```diff
@@ -152,12 +153,15 @@ func (c *AuthConfig) InstallAPI(container *restful.Container) []string {
 	osOAuthClientConfig.RedirectUrl = c.Options.MasterPublicURL + path.Join(OpenShiftOAuthAPIPrefix, tokenrequest.DisplayTokenEndpoint)
 
 	osOAuthClient, _ := osincli.NewClient(osOAuthClientConfig)
-	if c.MasterRoots != nil {
-		// Copy the default transport
-		var transport http.Transport = *http.DefaultTransport.(*http.Transport)
-		// Set TLS CA roots
-		transport.TLSClientConfig = &tls.Config{RootCAs: c.MasterRoots}
-		osOAuthClient.Transport = &transport
+	if c.Options.MasterCA != nil && len(*c.Options.MasterCA) > 0 {
```
Comment (deads2k) [inline pkg/cmd/server/origin/auth.go]: unnecessary nil check

**Comment by deads2k** [comment]
minor comment.

lgtm.

**Comment by liggitt** [comment]
[merge]

**Comment by liggitt** [comment]
unrelated:
```
!!! Error in hack/../test/end-to-end/../../hack/util.sh:220
	'return 1' exited with status 1
Call stack:
	1: hack/../test/end-to-end/../../hack/util.sh:220 wait_for_registry(...)
	2: hack/../test/end-to-end/core.sh:73 main(...)
```
[merge]
